File: Javascript DSA/Trees/Range Sum of BST.js
// Problem Link: https://leetcode.com/problems/range-sum-of-bst/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} low
 * @param {number} high
 * @return {number}
 */
var rangeSumBST = (root, low, high) => {
    
    if(!root) 
        return 0;
    let sum = ( root.val >= low && root.val <= high ? root.val : 0 );
    if( root.val >= low )
        sum += rangeSumBST( root.left, low, high);
    if( root.val <= high )
        sum += rangeSumBST( root.right, low, high);
    return sum;
};

File: Trees/Range Sum of BST.cpp
//Link: https://leetcode.com/problems/range-sum-of-bst/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// Using Inorder Traversal
// Using Recursion
// Time Complexity: O(N), where N is the number of nodes in the tree.
// Space Complexity: O(H), where H is the height of the tree. 

class Solution {
public:
    int sum = 0;
    int rangeSumBST(TreeNode* root, int low, int high) {
        if( !root ) return 0;
        if( root -> left != NULL)
            rangeSumBST( root-> left, low, high);
        if( root-> val >= low && root -> val <= high)
        {
            sum+= root-> val;
        }
        if( root -> right != NULL)
        {
            rangeSumBST( root-> right, low, high);
        }
        return sum;
    }
};
